### PR TITLE
[STEP-192]: Export all app dSYMs into the zip

### PR DESCRIPTION
**Pull request by istvankovacs-bitrise**
https://bitrise.atlassian.net/browse/STEP-192

### What?
All of the app `dSYM`s are exported into the zip.

### Why?
Previously only one app `dSYM` was exported, but since app clip has its own app `dSYM` generated, it has to be exported as well.

### How?
The `dSYM` collecting logic has been updated in the `go-xcode` library by [this](https://github.com/bitrise-io/go-xcode/pull/95) PR. This Step needed to be modified to export all of the `dSYM`s returned by the util function mentioned above.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -703,20 +711,14 @@ is available in the $BITRISE_IDEDISTRIBUTION_LOGS_PATH environment variable`)
 			fail("Failed to create tmp dir, error: %s", err)
 		}
 
-		if appDSYM != "" {
-			if err := command.CopyDir(appDSYM, dsymDir, false); err != nil {
-				fail("Failed to copy (%s) -> (%s), error: %s", appDSYM, dsymDir, err)
-			}
+		if len(appDSYM) > 0 {
+			exportDSYMs(dsymDir, appDSYM)
```
Comment (istvankovacs-bitrise) [inline main.go:715]: After `appDSYM` made to be an array, the exporting algorithm turned to be the same as below, so it made sense to extract it.

Comment (istvankovacs-bitrise) [reply main.go:715]: You may find the related change in `go-xcode` [here](https://github.com/bitrise-io/go-xcode/pull/95).

File: main.go [hunk 2]
```diff
@@ -130,6 +130,14 @@ func determineExportMethod(desiredExportMethod string, archiveExportMethod expor
 	return exportMethod, nil
 }
 
+func exportDSYMs(dsymDir string, dsyms []string) {
+	for _, dsym := range dsyms {
+		if err := command.CopyDir(dsym, dsymDir, false); err != nil {
+			fail("Failed to copy (%s) -> (%s), error: %s", dsym, dsymDir, err)
```
Comment (lpusok) [inline main.go:136]: Please do not call os.Exit in a function other than main.

File: main.go [hunk 3]
```diff
@@ -130,6 +130,15 @@ func determineExportMethod(desiredExportMethod string, archiveExportMethod expor
 	return exportMethod, nil
 }
 
+func exportDSYMs(dsymDir string, dsyms []string) error {
+	for _, dsym := range dsyms {
+		if err := command.CopyDir(dsym, dsymDir, false); err != nil {
+			return fmt.Errorf("Failed to copy (%s) -> (%s), error: %s", dsym, dsymDir, err)
```
Comment (lpusok) [inline main.go:136]: Error strings should be lowercase and not contain "." as per golang standards. Additionally, "something, error: reason" is not a recommended pattern, use "sonething: reason" instead, as seen in the Wrap function (https://godoc.org/github.com/pkg/errors#Wrap)
```suggestion
			return fmt.Errorf("failed to copy (%s) -> (%s): %s", dsym, dsymDir, err)
```

File: main.go [hunk 4]
```diff
@@ -703,19 +712,17 @@ is available in the $BITRISE_IDEDISTRIBUTION_LOGS_PATH environment variable`)
 			fail("Failed to create tmp dir, error: %s", err)
 		}
 
-		if appDSYM != "" {
-			if err := command.CopyDir(appDSYM, dsymDir, false); err != nil {
-				fail("Failed to copy (%s) -> (%s), error: %s", appDSYM, dsymDir, err)
+		if len(appDSYM) > 0 {
+			if err := exportDSYMs(dsymDir, appDSYM); err != nil {
+				fail(err.Error())
```
Comment (lpusok) [inline main.go:717]:
```suggestion
				fail("Failed to export DSYMs: %v", err)
```

File: step.yml [hunk 5]
```diff
@@ -344,14 +344,14 @@ outputs:
     opts:
       title: The created .dSYM dir's path
       description: |-
-        If `export_all_dsyms` is set to `"yes"`, step will collect every dsym (.app dsym and framwork dsyms) in a directory and export the directory path.
-        Otherwise this environment points to the .app.DSYM directory path.
+        This environment points to the path of the directory which contains the dSYMs files.
```
Comment (BanyikAnna) [inline step.yml:347]: This Environment Variable points to the path of the directory which contains the dSYMs files.

File: main.go (hunk 3 above)
Comment (istvankovacs-bitrise) [reply main.go:136]: Many thanks for the clarification and the detailed answer!

File: step.yml [hunk 6]
```diff
@@ -344,14 +344,14 @@ outputs:
...
+        If `export_all_dsyms` is set to `"yes"`, step will collect every dSYM (app dSYMs and framwork dSYMs).
```
Comment (BanyikAnna) [inline step.yml:348]: If `export_all_dsyms` is set to `yes`, the Step will collect every dSYM (app dSYMs and framwork dSYMs).

File: step.yml [hunk 7]
```diff
@@ -344,14 +344,14 @@ outputs:
...
   - BITRISE_DSYM_PATH:
     opts:
       title: The created .dSYM.zip file's path
       description: |-
-        If `export_all_dsyms` is set to `"yes"`, step will collect every dsym (.app dsym and framwork dsyms) in a directory, zip it and export the zipped directory path.
-        Otherwise this environment points to the zipped .app.DSYM directory path.
+        This environment points to the path of the zip file which contains the dSYM files.
```
Comment (BanyikAnna) [inline step.yml:353]: This Environment Variable points to the path of the zip file which contains the dSYM files.

File: step.yml [hunk 8]
```diff
@@ -344,14 +344,14 @@ outputs:
...
+        If `export_all_dsyms` is set to `"yes"`, step will collect every dSYM (app dSYMs and framwork dSYMs).
```
Comment (BanyikAnna) [inline step.yml:354]: If `export_all_dsyms` is set to `yes`, the Step will collect every dSYM (app dSYMs and framwork dSYMs).

File: step.yml (hunk 6 above)
Comment (lpusok) [inline step.yml:348]:
```suggestion
        If `export_all_dsyms` is set to `yes`, the Step will also collect framework dSYMs in addition to app dSYMs.
```

Comment (istvankovacs-bitrise) [reply step.yml:348]: Thanks for the suggestion, I will update these based on Anna's comments.

File: main.go [hunk 9]
```diff
@@ -703,19 +712,17 @@ is available in the $BITRISE_IDEDISTRIBUTION_LOGS_PATH environment variable`)
...
 			log.Warnf("no app dsyms found")
 		}
 
 		if cfg.ExportAllDsyms {
-			for _, dsym := range frameworkDSYMs {
-				if err := command.CopyDir(dsym, dsymDir, false); err != nil {
-					fail("Failed to copy (%s) -> (%s), error: %s", dsym, dsymDir, err)
-				}
+			if err := exportDSYMs(dsymDir, frameworkDSYMs); err != nil {
+				fail(err.Error())
```
Comment (lpusok) [inline main.go:725]:
```suggestion
				fail("Failed to export dSYMs: %v", err)
```

File: main.go [hunk 10]
```diff
@@ -703,19 +712,17 @@ is available in the $BITRISE_IDEDISTRIBUTION_LOGS_PATH environment variable`)
 			fail("Failed to create tmp dir, error: %s", err)
 		}
 
-		if appDSYM != "" {
-			if err := command.CopyDir(appDSYM, dsymDir, false); err != nil {
-				fail("Failed to copy (%s) -> (%s), error: %s", appDSYM, dsymDir, err)
+		if len(appDSYM) > 0 {
+			if err := exportDSYMs(dsymDir, appDSYM); err != nil {
+				fail("Failed to export DSYMs: %v", err)
```
Comment (lpusok) [inline main.go:717]:
```suggestion
				fail("Failed to export dSYMs: %v", err)
```

**Review by lpusok — APPROVED** [review]

**Comment by istvankovacs-bitrise** [comment]
@lpusok, many thanks for the thorough out review! Would you mind merging it, please?

**Pull request merged**